[#1002, #1543, #1677] Add hooks to `Item5e#roll` & `Item5e#displayCard` to allow modules to enhance that workflow

arbron opened the pull request:
Adds a number of hooks to the item usage process to allow modules to better customize that workflow without having to override or wrap system methods:

- `dnd5e.preUseItem`: Called before the item is configured, allows changing configuration data, controlling whether the dialog should be displayed, and upcasting spells even if the dialog isn't shown
- `dnd5e.preItemUsageConsumption`: Called before usage updates are calculated
- `dnd5e.itemUsageConsumption`: Called after resource consumption is calculated to allow any modifications before those changes are applied to the actor or other items
- `dnd5e.preDisplayCard`: Called before the chat card is created with the chat data that will be used to display the card to allow for any changes before the creation occurs
- `dnd5e.displayCard`: Called after the chat card is created to allow for any changes now that the card is saved to the database
- `dnd5e.useItem`: Called after consumption updates, the chat card is created, and any measured template is created to respond to the effects of the usage (e.g. auto applying damage to all creatures within the template)

This also makes the process for placing a measured template async so that it can be awaited and sets the creating item's UUID as a flag so their origin can be tracked.

aaclayton commented on module/documents/item.mjs:
```diff
@@ -642,101 +642,169 @@ export default class Item5e extends Item {
+     * @param {ItemRollOptions} options          Additional options used for configuring item rolls.
+     * @returns {boolean}                        Explicitly return false to prevent item from being rolled.
+     */
+    if ( Hooks.call("dnd5e.preRoll", item, config, options) === false ) return;
```
I think this hook name is possibly too general. The method here is effectively "play an item card to chat" - the method is called `roll` because in early versions of `dnd5e` that is typically what resulted from this workflow.... but a side comment is that perhaps `roll` is not the right name for this method either.

Regardless, I think creating hooks for `dnd5e.preRoll` or `dnd5e.roll` are too easily confused with functionality that might apply at the time that an actual dice roll occurs rather than as part of the item chat card.

arbron replied:
How about `dnd5e.preUseItem` & `dnd5e.useItem` (and other related hook changes) to match the term of `_getUsageUpdates`?

aaclayton reviewed (changes requested):
Looks good, thank you @arbron - I would like to see us tighten up our naming conventions a bit with hooks and methods. 

Also, as a separate request: it would be helpful for some topics which are conceptually distinct to be separate MRs. The AbilityTemplate changes could have been reviewed and merged separately from the item chat card rolling stuff.

arbron commented:
@aaclayton With the new hook names, should I also create a new preferred `Item5e#use` method and indicate it should be used over `roll`?

aaclayton commented:
> @aaclayton With the new hook names, should I also create a new preferred `Item5e#use` method and indicate it should be used over `roll`?

I'm open to your preferences here, but I do think there is a good argument in favor of migrating to `Item5e#use` with a backwards-compatible deprecation for `Item5e#roll`

aaclayton commented on less/actors.less:
```diff
@@ -501,10 +501,10 @@
       .item-name {
         cursor: pointer;
-        &.rollable:hover .item-image {
+        &:is(.rollable, .usable):hover .item-image {
```
Can we move to using `.usable` directly rather than needing to support both `.rollable` and `.usable`? OR can we leave CSS alone here and just keep using `.rollable` and make this just an JS API change for now?

arbron replied:
I'll just leave the CSS alone for now.

arbron commented:
> I'm open to your preferences here, but I do think there is a good argument in favor of migrating to `Item5e#use` with a backwards-compatible deprecation for `Item5e#roll`

Renamed and deprecated. I also adjusted the css class from `rollable` to `usable`, but left backwards compatibility so we don't break 3rd party sheets.

aaclayton reviewed (changes requested):
When I suggested switching to `Item5e#use` I was envisioning a JS-only change. Seeing all the CSS changes makes me immediately regret that and want to switch back. Can we avoid having `:is(.rollable, .usable)` everywhere which is ugly.

aaclayton commented:
> Renamed and deprecated. I also adjusted the css class from `rollable` to `usable`, but left backwards compatibility so we don't break 3rd party sheets.

We are *very likely* already breaking 3rd party sheets so I would prefer to lean in rather than just volunteer to take on future tech debt.

aaclayton reviewed (approved):
Thank you!
